zebra-chain/src/transaction/serialize/sapling_v6.rs

```diff
@@ -0,0 +1,250 @@
+    // See [`redjubjub::Signature<SpendAuth>::zcash_deserialize`].
+    let spend_sigs: Vec<VersionedSig<Signature<SpendAuth>>> =
+        zcash_deserialize_external_count(spends_count, &mut reader)?;
+
+    // Denoted as `vOutputProofsSapling` in the spec.
+    //
+    // # Consensus
+    //
+    // > Elements of an Output description MUST be valid encodings of the types given above.
+    //
+    // https://zips.z.cash/protocol/protocol.pdf#outputdesc
+    //
+    // Type is `ZKOutput.Proof`, described in
+    // https://zips.z.cash/protocol/protocol.pdf#grothencoding
+    // It is not enforced here; this just reads 192 bytes.
+    // The type is validated when validating the proof, see
+    // [`groth16::Item::try_from`]. In #3179 we plan to validate here instead.
+    let output_proofs = zcash_deserialize_external_count(outputs_count, &mut reader)?;
+
+    // Denoted as `bindingSigSapling` in the spec.
+    let binding_sig = VersionedSig::zcash_deserialize(&mut reader)?.as_tx_v6()?;
+
+    // Create shielded spends from deserialized parts
+    let spends: Vec<_> = spend_prefixes
+        .into_iter()
+        .zip(spend_proofs)
+        .zip(spend_sigs)
+        .map(|((prefix, proof), spend_sig)| {
+            spend_sig.as_tx_v6().map(|sig| {
+                sapling::Spend::<sapling::SharedAnchor>::from_v5_parts(prefix, proof, sig)
+            })
+        })
+        .collect::<Result<Vec<_>, _>>()?;
+
+    // Create shielded outputs from deserialized parts
+    let outputs = output_prefixes
+        .into_iter()
+        .zip(output_proofs)
+        .map(|(prefix, proof)| sapling::Output::from_v5_parts(prefix, proof))
+        .collect();
+
+    // Create transfers
+    //
+    // # Consensus
+    //
+    // > The anchor of each Spend description MUST refer to some earlier
+    // > block’s final Sapling treestate. The anchor is encoded separately
+    // > in each Spend description for v4 transactions, or encoded once and
+    // > shared between all Spend descriptions in a v5 transaction.
+    //
+    // <https://zips.z.cash/protocol/protocol.pdf#spendsandoutputs>
+    //
+    // This rule is also implemented in
+    // [`zebra_state::service::check::anchor`] and
+    // [`zebra_chain::sapling::spend`].
+    //
+    // The "anchor encoding for v5 transactions" is implemented here.
+    let transfers = match shared_anchor {
+        Some(shared_anchor) => sapling::TransferData::SpendsAndMaybeOutputs {
+            shared_anchor,
+            spends: spends
+                .try_into()
+                .expect("checked spends when parsing shared anchor"),
+            maybe_outputs: outputs,
+        },
+        None => sapling::TransferData::JustOutputs {
+            outputs: outputs
+                .try_into()
+                .expect("checked spends or outputs and returned early"),
+        },
+    };
+
+    Ok(Some(sapling::ShieldedData {
+        value_balance,
+        transfers,
+        binding_sig,
+    }))
+}
```

Comment:
Let's move the content of this file into `transaction/serialize.rs`(bottom) and reduce duplicated comments to minimum.

Reply:
Instead duplicated comments, please add a todo, to extract common logic after tx-v6 is merged.